src/main/scala/com/microsoft/hyperspace/index/rankers/JoinIndexRanker.scala

```diff
@@ -16,40 +16,78 @@
+   * for on-the-fly shuffle or merge.
    *
-   * @param indexPairs index pairs for left and right side of the join. All index pairs are
+   * @param spark SparkSession.
+   * @param leftChild Logical relation of left child of the join.
+   * @param rightChild Logical relation of right child of the join.
+   * @param indexPairs Index pairs for left and right side of the join. All index pairs are
    *                   compatible with each other.
-   * @return rearranged index pairs according to their ranking. The first is the best.
+   * @return Rearranged index pairs according to their ranking. The first is the best.
    */
   def rank(
+      spark: SparkSession,
+      leftChild: LogicalPlan,
+      rightChild: LogicalPlan,
       indexPairs: Seq[(IndexLogEntry, IndexLogEntry)]): Seq[(IndexLogEntry, IndexLogEntry)] = {
-
+    val hybridScanEnabled = HyperspaceConf.hybridScanEnabled(spark)
     indexPairs.sortWith {
-      case ((left1, left2), (right1, right2)) =>
-        if (left1.numBuckets == left2.numBuckets && right1.numBuckets == right2.numBuckets) {
-          left1.numBuckets > right1.numBuckets
-        } else if (left1.numBuckets == left2.numBuckets) {
+      case ((left1, right1), (left2, right2)) =>
+        // These common bytes were calculated and tagged in getCandidateIndexes.
+        // The value is the summation of common source files of the given plan and each index.
+        lazy val leftCommonBytes = left1
+          .getTagValue(leftChild, IndexLogEntryTags.COMMON_BYTES)
+          .getOrElse(0L) + right1
+          .getTagValue(rightChild, IndexLogEntryTags.COMMON_BYTES)
+          .getOrElse(0L)
+        lazy val rightCommonBytes = left2
+          .getTagValue(leftChild, IndexLogEntryTags.COMMON_BYTES)
+          .getOrElse(0L) + right2
+          .getTagValue(rightChild, IndexLogEntryTags.COMMON_BYTES)
+          .getOrElse(0L)
+
+        if (left1.numBuckets == right1.numBuckets && left2.numBuckets == right2.numBuckets) {
+          if (!hybridScanEnabled || (leftCommonBytes == rightCommonBytes)) {
+            left1.numBuckets > left2.numBuckets
+          } else {
+            // If both index pairs have the same number of buckets and Hybrid Scan is enabled,
+            // pick the pair with more common bytes with the given source plan, so as to
+            // reduce the overhead from handling appended and deleted files.
+            leftCommonBytes > rightCommonBytes
+          }
+        } else if (left1.numBuckets == right1.numBuckets) {
           true
-        } else if (right1.numBuckets == right2.numBuckets) {
+        } else if (left2.numBuckets == right2.numBuckets) {
           false
-        } else {
+        } else if (!hybridScanEnabled) {
           true
+        } else {
+          // Pick the pair with "more common bytes" if both pairs have different number of buckets.
+          leftCommonBytes > rightCommonBytes
```

Comment:
How about:
```scala
} else {
  // At this point, both pairs have different number of buckets. If Hybrid Scan is enabled, pick the pair
  // with "more common bytes", otherwise pick the first pair.
  !hybridScanEnabled || leftCommonBytes > rightCommonBytes
}
```